app/src/main/cpp/ExternalVR.cpp

```diff
@@ -367,11 +386,26 @@ ExternalVR::PushFramePoses(const vrb::Matrix& aHeadTransform, const std::vector<
     }
     immersiveController.hand = controller.leftHanded ? mozilla::gfx::ControllerHand::Left : mozilla::gfx::ControllerHand::Right;
 
-    immersiveController.flags = mozilla::gfx::ControllerCapabilityFlags::Cap_Orientation;
-    immersiveController.isOrientationValid = true;
-    vrb::Quaternion quaternion(controller.transformMatrix);
-    quaternion = quaternion.Inverse();
-    memcpy(&(immersiveController.pose.orientation), quaternion.Data(), sizeof(immersiveController.pose.orientation));
+    const uint16_t flags = GetControllerCapabilityFlags(controller.deviceCapabilities);
+    immersiveController.flags = static_cast<mozilla::gfx::ControllerCapabilityFlags>(flags);
+
+    if (flags & static_cast<uint16_t>(mozilla::gfx::ControllerCapabilityFlags::Cap_Orientation)) {
+      immersiveController.isOrientationValid = true;
+
+      vrb::Quaternion quaternion(controller.transformMatrix);
+      quaternion = quaternion.Inverse();
+      memcpy(&(immersiveController.pose.orientation), quaternion.Data(), sizeof(immersiveController.pose.orientation));
+    }
+    if (flags & static_cast<uint16_t>(mozilla::gfx::ControllerCapabilityFlags::Cap_Position)) {
+      immersiveController.isPositionValid = true;
+
+      // For 6 DOF controllers, we added this kAverageHeight in order to have the same height with 3 DOF controllers.
+      // So, we need to decrease it to provide the proper position data to the immersive controllers.
+      const vrb::Vector kAverageHeight(0.0f, 1.7f, 0.0f);
```

Comment:
For 6DOF controllers we have to minus kAverageHeight, it is because we add this kAverageHeight at, https://github.com/MozillaReality/FirefoxReality/blob/master/app/src/oculusvr/cpp/DeviceDelegateOculusVR.cpp#L654 to close the distance as 3 DOF. Besides, when using ExternalVR to PushFramePoses(), we are usually already at the RenderMode::Immersive. So, I don't understand if we still need to do the same for consistency.

Reply:
What I mean is adding this if in `DeviceDelegateOculusVR.cpp`

```
if (state[i].controllerCapabilities.ControllerCapabilities & ovrControllerCaps_HasPositionTracking) {
        auto & position = tracking.HeadPose.Pose.Position;
        state[i].controllerTransform.TranslateInPlace(vrb::Vector(position.x, position.y, position.z));
        if (m.renderMode == device::RenderMode::StandAlone) {
            head.TranslateInPlace(kAverageHeight);
        }
        flags |= device::Position;
}
```

That way we don't need to duplicate the constant in VRExternal. And the consistency because we are doing the same if for the head position kAverageHeight

Reply:
Thanks I have fixed it. By checking
```
 if (renderMode == device::RenderMode::StandAlone) {
          position.x += kAverageHeight.x();
          position.y += kAverageHeight.y();
          position.z += kAverageHeight.z();
 }
```
in `UpdateControllers()`.